python/dgl/distributed/sparse_emb.py

```diff
@@ -58,148 +59,147 @@ class DistEmbedding:
 
     Note
     ----
-    When a ``DistEmbedding``  object is used when the deep learning framework is recording
-    the forward computation, users have to invoke py:meth:`~dgl.distributed.SparseAdagrad.step`
-    afterwards. Otherwise, there will be some memory leak.
+    When a ``NodeEmbedding``  object is used when the deep learning framework is recording
+    the forward computation, users have to invoke
+    py:meth:`~dgl.distributed.optim.SparseAdagrad.step` afterwards. Otherwise, there will be
+    some memory leak.
     '''
     def __init__(self, num_embeddings, embedding_dim, name=None,
                  init_func=None, part_policy=None):
         self._tensor = DistTensor((num_embeddings, embedding_dim), F.float32, name,
-                                  init_func, part_policy)
+                                  init_func=init_func, part_policy=part_policy)
         self._trace = []
-
-    def __call__(self, idx):
+        self._name = name
+        self._num_embeddings = num_embeddings
+        self._embedding_dim = embedding_dim
+
+        # Check whether it is multi-gpu/distributed training or not
+        if th.distributed.is_initialized():
+            self._rank = th.distributed.get_rank()
+            self._world_size = th.distributed.get_world_size()
+        else:
+            assert 'th.distributed shoud be initialized'
+        self._optm_state = None # track optimizer state
+        self._part_policy = part_policy
+
+    def __call__(self, idx, device=th.device('cpu')):
+        """
+        node_ids : th.tensor
+            Index of the embeddings to collect.
+        device : th.device
+            Target device to put the collected embeddings.
+
+        Returns
+        -------
+        Tensor
+            The requested node embeddings
+        """
         idx = utils.toindex(idx).tousertensor()
-        emb = self._tensor[idx]
+        emb = self._tensor[idx].to(device, non_blocking=True)
         if F.is_recording():
             emb = F.attach_grad(emb)
-            self._trace.append((idx, emb))
+            self._trace.append((idx.to(device, non_blocking=True), emb))
```

Comment:
what was the decision for async read?

Reply:
The idx is not used right after this call. Only emb is used immediately. Thus  we can use async to here.

Reply:
this is still very tricky when used with the async sampling pipeline. We cannot perform any GPU operations in the sampling thread. otherwise, it'll interfere with GPU operations in the training thread. I guess we can only perform this operation in the training thread.

Reply:
There will be some other issue.
The gradient of emb will be moved to CPU by default in pytorch.